src/hotspot/share/utilities/hashtable.hpp

```diff
@@ -242,15 +242,35 @@ class KVHashtable : public BasicHashtable<F> {
     return (KVHashtableEntry*)BasicHashtable<F>::bucket(i);
   }
 
+  // The following method is not MT-safe and must be done under lock.
+  KVHashtableEntry** bucket_addr(int i) {
+    return (KVHashtableEntry**)BasicHashtable<F>::bucket_addr(i);
+  }
+
   KVHashtableEntry* new_entry(unsigned int hashValue, K key, V value) {
     KVHashtableEntry* entry = (KVHashtableEntry*)BasicHashtable<F>::new_entry(hashValue);
     entry->_key   = key;
     entry->_value = value;
     return entry;
   }
 
+  void free_entry(KVHashtableEntry* entry) {
+    BasicHashtable<F>::free_entry(entry);
+  }
+
 public:
   KVHashtable(int table_size) : BasicHashtable<F>(table_size, sizeof(KVHashtableEntry)) {}
+  ~KVHashtable() {
+    KVHashtableEntry* probe = NULL;
+    for (int index = 0; index < table_size(); index++) {
+      for (KVHashtableEntry** p = bucket_addr(index); *p != NULL; ) {
+        probe = *p;
+        *p = probe->next();
```

Comment:
Should we also call the destructor (something like `probe->_value.~V()` , not sure about the syntax). I.e., similar to `GrowableArray<E>`:

https://github.com/openjdk/jdk/blob/e0f6f70d3f9e748d2bc53f371beca487e9343d4a/src/hotspot/share/utilities/growableArray.hpp#L499-L501

This way, we can get rid of this code, and move `delete ref()` into `SourceObjInfo::~SourceObjInfo()`

https://github.com/openjdk/jdk/blob/e0f6f70d3f9e748d2bc53f371beca487e9343d4a/src/hotspot/share/cds/archiveBuilder.hpp#L180-L186

Reply:
I wanted to do that (and even had a version that did), but unfortunately for this code, adding a destructor to SrcObjRef destroys the ref pointer in this scope, so we'd need to have a copy constructor etc to keep the pointer alive.

https://github.com/openjdk/jdk/blob/e0f6f70d3f9e748d2bc53f371beca487e9343d4a/src/hotspot/share/cds/archiveBuilder.cpp#L464

So I thought we should wait until we replace this table to make it better.  I was going to add a comment here.

Reply:
Sounds good. The current code is a bit clumsy but there's no memory leak anymore. We should eventually move all functionalities of KVHashTable into ResourceHashtable so we don't have 2 different partially working hashtable types!

Reply:
Yes, I agree.  This src_obj_table will still need some surgery, since the replacement table should call destructors on all the elements so maybe the table shouldn't contain a copy of SourceObjectInfo.